File: problem_31.py
# Problem 31 (Coin sums)
# ======================
#
# In the United Kingdom the currency is made up of pound (£) and pence (p).
# There are eight coins in general circulation:
#
#    1p, 2p, 5p, 10p, 20p, 50p, £1 (100p), and £2 (200p).
#
# It is possible to make £2 in the following way:
#
#    1×£1 + 1×50p + 2×20p + 1×5p + 1×2p + 3×1p
#
# How many different ways can £2 be made using any number of coins?

# This is a classic dynamic programming problem.
# One recursive solution would be:
options = [1,2,5,10,20,50,100,200]

def recursive(target_amount):
    """Generate all permutations that add up to the target amount."""
    for coin in options:
        if coin > target_amount:
            continue
        elif coin == target_amount:
            yield (coin,)
        else: # coin < target_amount
            for rest in recursive(target_amount - coin):
                yield (coin,) + rest

# But this can be done in another way.
def dynamic(target_amount):
    """Calculate number of all permutations that add up to the target amount."""
    cache = {0: 1}

    for amount in range(1, target_amount+1):
        # Number of all known paths that lead to this point
        ways = sum(cache[amount - coin] for coin in options if (amount - coin) in cache)
        cache[amount] = ways

    return cache[target_amount]


# However, we need combinations, not permutations.
# I see no way to calculate the number of combinations
# without somehow going through all of them.
#
# For illustration, suppose we have only 2p and 3p coins.
# The combinations for amount N are:
#
#     combinations(N-2) ⋃ combinations(N-3)
#
# ...after extending each one with the proper coin, of course.
#
# The pseudocode for the general case is:
#
#   combinations[0] = [Ø]
#   for amount in [1,2,..200]:
#       combinations[N] = union(extend(combinations[N-coin], coin) for coin in coins)
#
def dynamic_combinations(target_amount):
    combinations_for = {0: [{}]}

    for amount in range(1, target_amount+1):
        union = set()
        for coin in options:
            if amount - coin >= 0:
                for c in combinations_for[amount - coin]:
                    # extend the combination
                    c = c.copy()
                    c[coin] = c.get(coin, 0) + 1
                    # convert into canonical representation ("tuplify" it)
                    # so it's hashable
                    c = tuple(sorted(c.items()))
                    # add to union
                    union.add(c)
        
        # "untuplify" to get back dicts
        combinations_for[amount] = [dict(tup) for tup in union]

    return len(combinations_for[target_amount])


if __name__ == "__main__":
    print(dynamic_combinations(200))

File: problem_31.cpp
// Problem 31 (Coin sums)
// ======================
//
// In the United Kingdom the currency is made up of pound (£) and pence (p).
// There are eight coins in general circulation:
//
//    1p, 2p, 5p, 10p, 20p, 50p, £1 (100p), and £2 (200p).
//
// It is possible to make £2 in the following way:
//
//    1×£1 + 1×50p + 2×20p + 1×5p + 1×2p + 3×1p
//
// How many different ways can £2 be made using any number of coins?

// This is a classic dynamic programming problem.
// If we interpret ways as "permutations", the solution is:
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <unordered_set>

#define N_COINS 8
int coins[N_COINS] = {1, 2, 5, 10, 20, 50, 100, 200};

long dynamic_permutations(int *coins, int num_coins, int target_amount) {
    long *cache = (long *)calloc(target_amount, sizeof(long));

    cache[0] = 1;

    for (int amount=1; amount <= target_amount; amount++) {
        // Number of all known paths that lead to this point
        long ways = 0;
        for (int i=0; i < num_coins; i++) {
            if (amount - coins[i] >= 0) {
                ways += cache[amount - coins[i]];
            }
        }
        cache[amount] = ways;
    }

    long result = cache[target_amount];
    free(cache);

    return result;
}


// However, we need combinations, not permutations.
// I see no way to calculate the number of combinations
// without somehow going through all of them.
//
// For illustration, suppose we have only 2p and 3p coins.
// The combinations for amount N are:
//
//     combinations(N-2) ⋃ combinations(N-3)
//
// ...after extending each one with the proper coin, of course.
//
//
// But, since this is C++, we first need to do some boilerplate ourselves.

// o===========================o
// |   HELPERS / BOILERPLATE   |
// o===========================o
class Combination {
    public:
    int counts[N_COINS];
    Combination() {
        memset(this->counts, 0, N_COINS*sizeof(int));
    };

    Combination(const Combination& other) {  // the copy constructor
        memcpy(this->counts, other.counts, N_COINS*sizeof(int));
    }

    bool operator==(const Combination& other) const {
        return 0 == memcmp(this->counts, other.counts, N_COINS*sizeof(int));
    }

    Combination& operator=(const Combination& other) {
        if (this != &other) {
            memcpy(this->counts, other.counts, N_COINS*sizeof(int));
        }
        return *this;
    }
};

namespace std {
    // Override std::hash so we don't have to provide it explicitly
    template <> struct hash<Combination> {
        size_t operator()(const Combination & c) const {
            static_assert(N_COINS == 8, "Update the hash function!");
            return (2 * c.counts[0] + 3 * c.counts[1] + 5 * c.counts[2] + 7 * c.counts[3]
                  + 11 * c.counts[4] + 13 * c.counts[5] + 17 * c.counts[6] + 19 * c.counts[7]);
        };
    };
}

// o=========================o
// |   THE ACTUAL FUNCTION   |
// o=========================o

long dynamic_combinations(int *coins, int num_coins, int target_amount)
{
    std::unordered_set<Combination> *combinations_for = new std::unordered_set<Combination>[1 + target_amount];

    // There's only one way to get 0, and that's using 0 coins.
    combinations_for[0].insert(Combination());

    // In general, the number of combinations for N is:
    //
    //    UNION       extended(combinations_for[N-c])
    //   (c ∈ coins)
    for (int amount=1; amount <= target_amount; amount++) {

        for (int i=0; i < num_coins; i++) {
            if (amount - coins[i] >= 0) {
                // Extend each combination with the extra coin
                for (Combination c : combinations_for[amount - coins[i]]) {
                    c.counts[i] += 1;
                    combinations_for[amount].insert(c);
                }
            }
        }
    }

    return combinations_for[target_amount].size();
}


int main() {
    assert(2614632708599399614 == dynamic_permutations(coins, N_COINS, 80));
    printf("%ld\n", dynamic_combinations(coins, N_COINS, 200));
    return 0;
}
